src/storage/transaction/TossNoVerEdgeIterator.h

```diff
@@ -0,0 +1,203 @@
+            ttlDuration_ = ttl->value().second;
+        }
+        if (!checkIter()) {
+            next();
+        }
+    }
+
+    folly::StringPiece key() const override {
+        return iter_->valid() ? iter_->key() : lockIter_->key();
+    }
+
+    bool valid() const override {
+        return iter_->valid() || lockIter_->valid();
+    }
+
+    void next() override {
+        if (stopSearching_) {
+            return;
+        }
+
+        if (iter_->valid()) {
+            for (iter_->next(); iter_->valid(); iter_->next()) {
+                if (checkIter()) {
+                    return;
+                }
+            }
+        }
+
+        if (!iter_->valid()) {
+            if (!lockIter_) {
+                if (!resumeTasks_.empty()) {
+                    folly::collectAll(resumeTasks_).wait();
+                }
+                lockIter_ = std::make_unique<PendingLockIter>(
+                    pendingLocks_, [&](auto val) { return checkValue(val); });
+            } else {
+                if (lockIter_->valid()) {
+                    lockIter_->next();
+                }
+            }
+        }
+    }
+
+    // check if iter_ point to a valid edge key,
+    // and if it points to a lock, try to resume it.
+    bool checkIter() {
+        if (NebulaKeyUtils::isLock(planContext_->vIdLen_, iter_->key())) {
+            LOG_IF(INFO, FLAGS_trace_toss) << "meet a lock: " << folly::hexlify(iter_->key());
+            pendingLocks_.emplace_back(std::make_shared<PendingLock>(iter_->key()));
+            resumeTasks_.emplace_back(
+                planContext_->env_->txnMan_->resumeLock(planContext_, pendingLocks_.back()));
+            checkEdgeHasLock_ = true;
+        } else if (NebulaKeyUtils::isEdge(planContext_->vIdLen_, iter_->key())) {
+            LOG_IF(INFO, FLAGS_trace_toss) << "meet an edge: " << folly::hexlify(iter_->key());
+            if (hasPendingLock(iter_->key())) {
+                pendingLocks_.back()->edgeProps = iter_->val().str();
+                return false;
+            }
+            if (check()) {
+                return true;
+            }
+        } else {
+            LOG(ERROR) << "meet a weird key: " << folly::hexlify(iter_->key());
+        }
+        return false;
+    }
+
+    bool checkValue(folly::StringPiece val) {
+        reader_.reset(*schemas_, val);
+        if (!reader_) {
+            planContext_->resultStat_ = ResultStatus::ILLEGAL_DATA;
+            return false;
+        }
+
+        if (hasTtl_ && CommonUtils::checkDataExpiredForTTL(
+                           schemas_->back().get(), reader_.get(), ttlCol_, ttlDuration_)) {
+            reader_.reset();
+            return false;
+        }
+        return true;
+    }
+
+    bool hasPendingLock(const folly::StringPiece& ekey) {
+        if (!checkEdgeHasLock_) {
+            return false;
+        }
+        checkEdgeHasLock_ = false;
+
+        auto& prevLock = pendingLocks_.back()->lockKey;
```

Comment:
return false when `pendingLocks_` is empty?

Reply:
not necessary, the checkEdgeHasLock_ will guarantee pendingLocks_ is not empty.